### PR TITLE
Switch GitHub actions to actions-rust-lang/setup-toolchain

### PR DESCRIPTION
**Pull request by Twey**
## Motivation

`actions-rs` has a very old version of `rustup` and requires a legacy `rust-toolchain` file.  It has also been [recently archived](https://github.com/actions-rs/toolchain) for reasons unknown to the community.

## Proposal

This PR replaces it with the hopefully-compatible [`actions-rust-lang/setup-rust-toolchain`](https://github.com/actions-rust-lang/setup-rust-toolchain).

## Test Plan

If GitHub checks pass on this PR then the actions are correct.

## Release Plan

- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.
- This PR is adding or removing Cargo features.
- Release is blocked and/or tracked by other issues (see links below)

## Links

- https://github.com/actions-rs/toolchain/issues/216
- https://github.com/actions-rust-lang/setup-rust-toolchain
- https://twitter.com/robjtede/status/1714402698001363427
- https://twitter.com/fasterthanlime/status/1714389848235164035
- https://www.reddit.com/r/rust/comments/z1mlls/actionsrs_github_actions_need_more_maintainers_or/
- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: .github/workflows/rust.yml
Comment (Twey) [reply .github/workflows/rust.yml:134]: Okay, I'll submit a patch to the action — currently if you have a toolchain file there's no way to use a different toolchain than the one specified there.

**Comment by Twey** [comment]
At the moment this PR uses my branch on https://github.com/Twey/setup-rust-toolchain to allow for overriding the toolchain just for linting.  I have a PR open against https://github.com/actions-rust-lang/setup-rust-toolchain that will hopefully eventually be merged to enable this functionality for that upstream repository; I'm unclear whether we want to merge now (and revert the commit once the upstream repository includes the functionality) or wait until that is merged (and remove the commit pointing us at my own GitHub from this PR).

**Review by ma2bd — APPROVED** [review]

**Pull request merged**
